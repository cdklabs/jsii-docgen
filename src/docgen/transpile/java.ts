import * as reflect from 'jsii-reflect';
import * as transpile from './transpile';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const Case = require('case');

// Helper methods
const toCamelCase = (text?: string) => {
  return Case.camel(text ?? '');
};

const toUpperCamelCase = (test?: string) => {
  const camelCase = toCamelCase(test);
  return camelCase.charAt(0).toUpperCase() + camelCase.slice(1);
};

// [1, 2, 3] -> [[], [1], [1, 2], [1, 2, 3]]
const prefixArrays = <T>(arr: T[]): T[][] => {
  const out: T[][] = [[]];
  const prefix: T[] = [];
  for (const elem of arr) {
    prefix.push(elem);
    out.push([...prefix]);
  }
  return out;
};

/**
 * Hack to convert a jsii property to a parameter for
 * parameter expansion.
 */
const propertyToParameter = (
  callable: reflect.Callable,
  property: reflect.Property,
): reflect.Parameter => {
  return {
    docs: property.docs,
    method: callable,
    name: property.name,
    optional: property.optional,
    parentType: property.parentType,
    spec: property.spec,
    system: property.system,
    type: property.type,
    variadic: false,
  };
};

export class JavaTranspile extends transpile.TranspileBase {
  constructor() {
    super(transpile.Language.JAVA);
  }

  public moduleLike(
    moduleLike: reflect.ModuleLike,
  ): transpile.TranspiledModuleLike {
    const fqn: string = moduleLike.targets?.java?.package;
    if (!fqn) {
      throw new Error(
        `Java is not a supported target for module: ${moduleLike.fqn}`,
      );
    }

    if (moduleLike instanceof reflect.Submodule) {
      const parent = this.getParentModule(moduleLike);
      if (!parent) {
        throw new Error(`Could not find parent module of ${moduleLike.fqn}`);
      }
      const parentFqn = parent.targets?.java?.package;
      if (!fqn.startsWith(parentFqn)) {
        throw new Error(`Expected submodule ${fqn} to start with ${parentFqn} since it is its parent module.`);
      }
      // { name: "software.amazon.awscdk", submodule: "services.ecr" }
      return { name: parentFqn, submodule: fqn.substring(parentFqn.length + 1) };
    }

    return { name: fqn };
  }

  private getParentModule(moduleLike: reflect.ModuleLike): reflect.Assembly | undefined {
    const ts = moduleLike.system;
    for (const assembly of ts.assemblies) {
      const child = assembly.submodules.find((mod) => mod.fqn === moduleLike.fqn);
      if (child) {
        return assembly;
      }
    }
    return undefined;
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);

    const parameters = callable.parameters.sort(this.optionalityCompare);

    const requiredParams = parameters.filter((p) => !p.optional);
    const optionalParams = parameters.filter((p) => p.optional);

    const name = callable.name;

    // simulate Java method overloading
    const inputLists = prefixArrays(optionalParams).map((optionals) => {
      return [...requiredParams, ...optionals].map((p) => this.formatParameter(this.parameter(p)));
    });

    const signatures = inputLists.map((inputs) => {
      return this.formatSignature(name, inputs);
    });

    let invocations;

    if (this.isClassBuilderGenerated(callable)) {
      const struct = this.extractFirstStruct(parameters);

      // render using Java builder syntax (show no overloads)
      invocations = [this.formatClassBuilder(type, parameters, struct)];

      // flatten out the parameters so the user doesn't have to jump between
      // docs of Foo and FooProps
      for (const property of struct.allProperties) {
        const parameter = propertyToParameter(callable, property);
        parameters.push(parameter);
      }
    } else {
      invocations = callable.kind === reflect.MemberKind.Initializer
        // render with `new Class` syntax (showing all constructor overloads)
        ? inputLists.map((inputs) => this.formatClassInitialization(type, inputs))
        // render invocation as method calls (showing all method overloads)
        : inputLists.map((inputs) => this.formatInvocation(type, inputs, name));
    }

    return {
      name,
      parentType: type,
      import: this.formatImport(type),
      parameters,
      signatures,
      invocations,
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const indent = ' '.repeat(4);
    const inputs = struct.allProperties.map((p) =>
      this.formatBuilderMethod(this.property(p), indent),
    ).flat();
    return {
      type: type,
      name: struct.name,
      import: this.formatImport(type),
      initialization: this.formatStructBuilder(type, inputs),
    };
  }

  public interface(
    iface: reflect.InterfaceType,
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  public parameter(
    parameter: reflect.Parameter,
  ): transpile.TranspiledParameter {
    const typeRef = this.typeReference(parameter.type);
    return {
      name: parameter.name,
      parentType: this.type(parameter.parentType),
      typeReference: typeRef,
      optional: parameter.optional,
      declaration: this.formatProperty(parameter.name, typeRef),
    };
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    const typeRef = this.typeReference(property.type);
    return {
      name: property.name,
      parentType: this.type(property.parentType),
      typeReference: typeRef,
      optional: property.optional,
      declaration: this.formatProperty(property.name, typeRef),
    };
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }

  public unionOf(types: string[]): string {
    return types.join(' OR ');
  }

  public listOf(type: string): string {
    return `java.util.List<${type}>`;
  }

  public mapOf(type: string): string {
    return `java.util.Map<java.lang.String, ${type}>`;
  }

  public any(): string {
    return 'java.lang.Object';
  }

  public str(): string {
    return 'java.lang.String';
  }

  public number(): string {
    return 'java.lang.Number';
  }

  public boolean(): string {
    return 'java.lang.Boolean';
  }

  public json(): string {
    return 'com.fasterxml.jackson.databind.node.ObjectNode';
  }

  public date(): string {
    return 'java.time.Instant';
  }

  public readme(readme: string): string {
    return readme;
  }

  private formatImport(type: transpile.TranspiledType): string {
    return `import ${type.fqn};`;
  };

  private formatParameter(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty,
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${tf} ${transpiled.name}`;
  }

  private formatInputs(inputs: string[]): string {
    return inputs.join(', ');
  };

  private formatStructBuilder(type: transpile.TranspiledType, methods: string[]): string {
    const builder = `${type.name}.builder()`;
    return [
      builder,
      ...methods,
      '    .build();',
    ].join('\n');
  };

  private formatClassInitialization(
    type: transpile.TranspiledType,
    inputs: string[],
  ): string {
    return `new ${type.name}(${this.formatInputs(inputs)});`;
  };

  private formatClassBuilder(
    type: transpile.TranspiledType,
    parameters: reflect.Parameter[],
    struct: reflect.InterfaceType,
  ): string {
    const createArgs = this.formatInputs(parameters.map((p) => this.formatParameter(this.parameter(p))));
    const indent = ' '.repeat(4);
    const methods: string[] = struct.allProperties.map((p) =>
      this.formatBuilderMethod(this.property(p), indent),
    ).flat();
    return [
      `${type.name}.Builder.create(${createArgs})`,
      ...methods,
      `${indent}.build();`,
    ].join('\n');
  };


  private formatSignature(name: string, inputs: string[]) {
    return `public ${name}(${this.formatInputs(inputs)})`;
  };

  private formatBuilderMethod(
    transpiled: transpile.TranspiledParameter,
    indent: string,
  ): string[] {
    if (transpiled.optional) indent = '//' + indent.slice(2);
    const lowerCamel = toCamelCase(transpiled.name);
    const base = `${indent}.${lowerCamel}`;
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    // allow rendering union types as multiple overrided builder methods
    if (tf.includes(' OR ')) {
      const choices = tf.split(' OR ');
      return choices.map((typ) => `${base}(${typ})`);
    } else {
      return [`${base}(${tf})`];
    }
  }

  private formatInvocation(
    type: transpile.TranspiledType,
    inputs: string[],
    method: string,
  ): string {
    let target = type.submodule ? `${type.namespace}.${type.name}` : type.name;
    if (method) {
      target = `${target}.${method}`;
    }
    return `${target}(${this.formatInputs(inputs)})`;
  };

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter,
  ): number {
    if (!p1.optional && p2.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private isStruct(p: reflect.Parameter): boolean {
    return p.type.fqn ? p.system.findFqn(p.type.fqn).isDataType() : false;
  }

  private isClassBuilderGenerated(
    callable: reflect.Callable,
  ): boolean {
    if (callable.kind !== reflect.MemberKind.Initializer) {
      return false;
    }

    const parameters: reflect.Parameter[] = callable.parameters.sort(this.optionalityCompare);
    const firstStruct = parameters.find((param) => this.isStruct(param));

    // no builder is generated if there is no struct parameter
    if (!firstStruct) {
      return false;
    }

    return true;
  };

  /**
   * Extracts the first struct out of a list of parameters (and throws
   * if there is none), removing it from the array.
   */
  private extractFirstStruct(
    parameters: reflect.Parameter[],
  ): reflect.InterfaceType {
    const firstStruct = parameters.find((param) => this.isStruct(param));
    if (!firstStruct) {
      throw new Error('No struct found in parameter list.');
    }
    const struct = firstStruct.parentType.system.findInterface(firstStruct.type.fqn!);
    parameters.splice(parameters.indexOf(firstStruct), 1);
    return struct;
  }

  private formatProperty(
    name: string,
    typeReference: transpile.TranspiledTypeReference,
  ): string {
    const tf = typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    if (tf.includes(' OR ')) {
      return `public java.lang.Object get${toUpperCamelCase(name)}();`;
    } else {
      return `public ${tf} get${toUpperCamelCase(name)}();`;
    }
  }
}
