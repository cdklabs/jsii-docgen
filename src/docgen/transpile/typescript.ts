import * as reflect from 'jsii-reflect';
import * as transpile from './transpile';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const Case = require('case');

// Helpers
const toCamelCase = (text?: string) => {
  return Case.camel(text ?? '');
};

const formatArguments = (inputs: string[]) => {
  return inputs.join(', ');
};

const formatStructInitialization = (type: transpile.TranspiledType) => {
  const target = type.submodule ? `${type.namespace}.${type.name}` : type.name;
  return `const ${toCamelCase(type.name)}: ${target} = { ... }`;
};

const formatClassInitialization = (
  type: transpile.TranspiledType,
  inputs: string[],
) => {
  const target = type.submodule ? `${type.namespace}.${type.name}` : type.name;
  return `new ${target}(${formatArguments(inputs)})`;
};

const formatInvocation = (
  type: transpile.TranspiledType,
  inputs: string[],
  method: string,
) => {
  let target = type.submodule ? `${type.namespace}.${type.name}` : type.name;
  if (method) {
    target = `${target}.${method}`;
  }
  return `${target}(${formatArguments(inputs)})`;
};

const formatImport = (type: transpile.TranspiledType) => {
  if (type.submodule) {
    return `import { ${type.submodule} } from '${type.module}'`;
  } else {
    return `import { ${type.name} } from '${type.module}'`;
  }
};

const formatSignature = (name: string, inputs: string[]) => {
  return `public ${name}(${formatArguments(inputs)})`;
};

/**
 * A TypeScript transpiler.
 */
export class TypeScriptTranspile extends transpile.TranspileBase {
  constructor() {
    super(transpile.Language.TYPESCRIPT);
  }

  public readme(readme: string): string {
    return readme;
  }

  public unionOf(types: string[]): string {
    return `${types.join(' | ')}`;
  }

  public listOf(type: string): string {
    return `${type}[]`;
  }

  public mapOf(type: string): string {
    return `{[ key: string ]: ${type}}`;
  }

  public any(): string {
    return 'any';
  }

  public boolean(): string {
    return 'boolean';
  }

  public str(): string {
    return 'string';
  }

  public number(): string {
    return 'number';
  }

  public date(): string {
    return 'Date';
  }

  public json(): string {
    return 'object';
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    return {
      name: property.name,
      parentType: this.type(property.parentType),
      typeReference: this.typeReference(property.type),
      optional: property.optional,
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public parameter(
    parameter: reflect.Parameter,
  ): transpile.TranspiledParameter {
    return {
      name: parameter.name,
      parentType: this.type(parameter.parentType),
      typeReference: this.typeReference(parameter.type),
      optional: parameter.optional,
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    return {
      type: type,
      name: struct.name,
      import: formatImport(type),
      initialization: formatStructInitialization(type),
    };
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);
    const parameters = callable.parameters.sort(this.optionalityCompare);
    const name = callable.name;
    const inputs = parameters.map((p) => this.formatParameters(this.parameter(p)));

    const invocation =
      callable.kind === reflect.MemberKind.Initializer
        ? formatClassInitialization(type, inputs)
        : formatInvocation(type, inputs, name);

    return {
      name,
      parentType: type,
      import: formatImport(type),
      parameters,
      signatures: [formatSignature(name, inputs)],
      invocations: [invocation],
    };
  }

  public interface(
    iface: reflect.InterfaceType,
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  public moduleLike(
    moduleLike: reflect.ModuleLike,
  ): transpile.TranspiledModuleLike {
    if (moduleLike instanceof reflect.Submodule) {
      const fqnParts = moduleLike.fqn.split('.');
      return { name: fqnParts[0], submodule: fqnParts[1] };
    }
    return { name: moduleLike.fqn };
  }

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter,
  ): number {
    if (!p1.optional && p2.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private formatParameters(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty,
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${transpiled.name}${transpiled.optional ? '?' : ''}: ${tf}`;
  }
}
