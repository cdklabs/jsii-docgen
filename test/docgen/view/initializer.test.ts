import * as reflect from 'jsii-reflect';
import { JavaTranspile } from '../../../src/docgen/transpile/java';
import { PythonTranspile } from '../../../src/docgen/transpile/python';
import { TranspiledType } from '../../../src/docgen/transpile/transpile';
import { TypeScriptTranspile } from '../../../src/docgen/transpile/typescript';
import { Initializer } from '../../../src/docgen/view/initializer';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

const findInitializer = (): reflect.Initializer => {
  for (const klass of assembly.system.classes) {
    if (klass.initializer) {
      return klass.initializer;
    }
  }
  throw new Error('Assembly does not contain an initializer');
};

describe('python', () => {
  const transpile = new PythonTranspile();
  test('snapshot', () => {
    const initializer = new Initializer(transpile, findInitializer(), (t: TranspiledType) => `#${t.fqn}`);
    expect(initializer.render().render()).toMatchSnapshot();
  });
});

describe('typescript', () => {
  const transpile = new TypeScriptTranspile();
  test('snapshot', () => {
    const initializer = new Initializer(transpile, findInitializer(), (t: TranspiledType) => `#${t.fqn}`);
    expect(initializer.render().render()).toMatchSnapshot();
  });
});

describe('java', () => {
  const transpile = new JavaTranspile();
  test('snapshot', () => {
    const initializer = new Initializer(transpile, findInitializer(), (t: TranspiledType) => `#${t.fqn}`);
    expect(initializer.render().render()).toMatchSnapshot();
  });
});
