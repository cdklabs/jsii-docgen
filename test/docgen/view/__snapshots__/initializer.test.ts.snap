// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp snapshot 1`] = `
" Initializers <a name=\\"Amazon.CDK.AWS.ECR.CfnPublicRepository.Initializer\\" id=\\"amazoncdkawsecrcfnpublicrepositoryinitializer\\"></a>

\`\`\`csharp
using Amazon.CDK.AWS.ECR;

new CfnPublicRepository(Construct Scope, string Id, CfnPublicRepositoryProps Props = null);
\`\`\`

| **Name** | **Type** | **Description** |
| --- | --- | --- |
| [\`Scope\`](#amazoncdkawsecrcfnpublicrepositoryparameterscope)<span title=\\"Required\\">*</span> | [\`Amazon.CDK.Construct\`](#Amazon.CDK.Construct) | scope in which this resource is defined. |
| [\`Id\`](#amazoncdkawsecrcfnpublicrepositoryparameterid)<span title=\\"Required\\">*</span> | \`string\` | scoped id of the resource. |
| [\`Props\`](#amazoncdkawsecrcfnpublicrepositoryparameterprops) | [\`Amazon.CDK.AWS.ECR.CfnPublicRepositoryProps\`](#Amazon.CDK.AWS.ECR.CfnPublicRepositoryProps) | resource properties. |

---

# \`Scope\`<sup>Required</sup> <a name=\\"Amazon.CDK.AWS.ECR.CfnPublicRepository.parameter.Scope\\" id=\\"amazoncdkawsecrcfnpublicrepositoryparameterscope\\"></a>

- *Type:* [\`Amazon.CDK.Construct\`](#Amazon.CDK.Construct)

scope in which this resource is defined.

---

# \`Id\`<sup>Required</sup> <a name=\\"Amazon.CDK.AWS.ECR.CfnPublicRepository.parameter.Id\\" id=\\"amazoncdkawsecrcfnpublicrepositoryparameterid\\"></a>

- *Type:* \`string\`

scoped id of the resource.

---

# \`Props\`<sup>Optional</sup> <a name=\\"Amazon.CDK.AWS.ECR.CfnPublicRepository.parameter.Props\\" id=\\"amazoncdkawsecrcfnpublicrepositoryparameterprops\\"></a>

- *Type:* [\`Amazon.CDK.AWS.ECR.CfnPublicRepositoryProps\`](#Amazon.CDK.AWS.ECR.CfnPublicRepositoryProps)

resource properties.

---
"
`;

exports[`java snapshot 1`] = `
" Initializers <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepository.Initializer\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositoryinitializer\\"></a>

\`\`\`java
import software.amazon.awscdk.services.ecr.CfnPublicRepository;

CfnPublicRepository.Builder.create(Construct scope, java.lang.String id)
//  .repositoryCatalogData(java.lang.Object)
//  .repositoryName(java.lang.String)
//  .repositoryPolicyText(java.lang.Object)
//  .tags(java.util.List<CfnTag>)
    .build();
\`\`\`

| **Name** | **Type** | **Description** |
| --- | --- | --- |
| [\`scope\`](#softwareamazonawscdkservicesecrcfnpublicrepositoryparameterscope)<span title=\\"Required\\">*</span> | [\`software.amazon.awscdk.core.Construct\`](#software.amazon.awscdk.core.Construct) | scope in which this resource is defined. |
| [\`id\`](#softwareamazonawscdkservicesecrcfnpublicrepositoryparameterid)<span title=\\"Required\\">*</span> | \`java.lang.String\` | scoped id of the resource. |
| [\`repositoryCatalogData\`](#softwareamazonawscdkservicesecrcfnpublicrepositorypropsparameterrepositorycatalogdata) | \`java.lang.Object\` | \`AWS::ECR::PublicRepository.RepositoryCatalogData\`. |
| [\`repositoryName\`](#softwareamazonawscdkservicesecrcfnpublicrepositorypropsparameterrepositoryname) | \`java.lang.String\` | \`AWS::ECR::PublicRepository.RepositoryName\`. |
| [\`repositoryPolicyText\`](#softwareamazonawscdkservicesecrcfnpublicrepositorypropsparameterrepositorypolicytext) | \`java.lang.Object\` | \`AWS::ECR::PublicRepository.RepositoryPolicyText\`. |
| [\`tags\`](#softwareamazonawscdkservicesecrcfnpublicrepositorypropsparametertags) | java.util.List<[\`software.amazon.awscdk.core.CfnTag\`](#software.amazon.awscdk.core.CfnTag)> | \`AWS::ECR::PublicRepository.Tags\`. |

---

# \`scope\`<sup>Required</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepository.parameter.scope\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositoryparameterscope\\"></a>

- *Type:* [\`software.amazon.awscdk.core.Construct\`](#software.amazon.awscdk.core.Construct)

scope in which this resource is defined.

---

# \`id\`<sup>Required</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepository.parameter.id\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositoryparameterid\\"></a>

- *Type:* \`java.lang.String\`

scoped id of the resource.

---

# \`repositoryCatalogData\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.parameter.repositoryCatalogData\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositorypropsparameterrepositorycatalogdata\\"></a>

- *Type:* \`java.lang.Object\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---

# \`repositoryName\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.parameter.repositoryName\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositorypropsparameterrepositoryname\\"></a>

- *Type:* \`java.lang.String\`

\`AWS::ECR::PublicRepository.RepositoryName\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname)

---

# \`repositoryPolicyText\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.parameter.repositoryPolicyText\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositorypropsparameterrepositorypolicytext\\"></a>

- *Type:* \`java.lang.Object\`

\`AWS::ECR::PublicRepository.RepositoryPolicyText\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext)

---

# \`tags\`<sup>Optional</sup> <a name=\\"software.amazon.awscdk.services.ecr.CfnPublicRepositoryProps.parameter.tags\\" id=\\"softwareamazonawscdkservicesecrcfnpublicrepositorypropsparametertags\\"></a>

- *Type:* java.util.List<[\`software.amazon.awscdk.core.CfnTag\`](#software.amazon.awscdk.core.CfnTag)>

\`AWS::ECR::PublicRepository.Tags\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags)

---
"
`;

exports[`python snapshot 1`] = `
" Initializers <a name=\\"aws_cdk.aws_ecr.CfnPublicRepository.Initializer\\" id=\\"awscdkawsecrcfnpublicrepositoryinitializer\\"></a>

\`\`\`python
import aws_cdk.aws_ecr

aws_cdk.aws_ecr.CfnPublicRepository(
  scope: Construct,
  id: str,
  repository_catalog_data: typing.Any = None,
  repository_name: str = None,
  repository_policy_text: typing.Any = None,
  tags: typing.List[CfnTag] = None
)
\`\`\`

| **Name** | **Type** | **Description** |
| --- | --- | --- |
| [\`scope\`](#awscdkawsecrcfnpublicrepositoryparameterscope)<span title=\\"Required\\">*</span> | [\`aws_cdk.core.Construct\`](#aws_cdk.core.Construct) | scope in which this resource is defined. |
| [\`id\`](#awscdkawsecrcfnpublicrepositoryparameterid)<span title=\\"Required\\">*</span> | \`str\` | scoped id of the resource. |
| [\`repository_catalog_data\`](#awscdkawsecrcfnpublicrepositorypropsparameterrepositorycatalogdata) | \`typing.Any\` | \`AWS::ECR::PublicRepository.RepositoryCatalogData\`. |
| [\`repository_name\`](#awscdkawsecrcfnpublicrepositorypropsparameterrepositoryname) | \`str\` | \`AWS::ECR::PublicRepository.RepositoryName\`. |
| [\`repository_policy_text\`](#awscdkawsecrcfnpublicrepositorypropsparameterrepositorypolicytext) | \`typing.Any\` | \`AWS::ECR::PublicRepository.RepositoryPolicyText\`. |
| [\`tags\`](#awscdkawsecrcfnpublicrepositorypropsparametertags) | typing.List[[\`aws_cdk.core.CfnTag\`](#aws_cdk.core.CfnTag)] | \`AWS::ECR::PublicRepository.Tags\`. |

---

# \`scope\`<sup>Required</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepository.parameter.scope\\" id=\\"awscdkawsecrcfnpublicrepositoryparameterscope\\"></a>

- *Type:* [\`aws_cdk.core.Construct\`](#aws_cdk.core.Construct)

scope in which this resource is defined.

---

# \`id\`<sup>Required</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepository.parameter.id\\" id=\\"awscdkawsecrcfnpublicrepositoryparameterid\\"></a>

- *Type:* \`str\`

scoped id of the resource.

---

# \`repository_catalog_data\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.parameter.repository_catalog_data\\" id=\\"awscdkawsecrcfnpublicrepositorypropsparameterrepositorycatalogdata\\"></a>

- *Type:* \`typing.Any\`

\`AWS::ECR::PublicRepository.RepositoryCatalogData\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorycatalogdata)

---

# \`repository_name\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.parameter.repository_name\\" id=\\"awscdkawsecrcfnpublicrepositorypropsparameterrepositoryname\\"></a>

- *Type:* \`str\`

\`AWS::ECR::PublicRepository.RepositoryName\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositoryname)

---

# \`repository_policy_text\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.parameter.repository_policy_text\\" id=\\"awscdkawsecrcfnpublicrepositorypropsparameterrepositorypolicytext\\"></a>

- *Type:* \`typing.Any\`

\`AWS::ECR::PublicRepository.RepositoryPolicyText\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-repositorypolicytext)

---

# \`tags\`<sup>Optional</sup> <a name=\\"aws_cdk.aws_ecr.CfnPublicRepositoryProps.parameter.tags\\" id=\\"awscdkawsecrcfnpublicrepositorypropsparametertags\\"></a>

- *Type:* typing.List[[\`aws_cdk.core.CfnTag\`](#aws_cdk.core.CfnTag)]

\`AWS::ECR::PublicRepository.Tags\`.

> [http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html#cfn-ecr-publicrepository-tags)

---
"
`;

exports[`typescript snapshot 1`] = `
" Initializers <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepository.Initializer\\" id=\\"awscdkawsecrcfnpublicrepositoryinitializer\\"></a>

\`\`\`typescript
import { CfnPublicRepository } from '@aws-cdk/aws-ecr'

new CfnPublicRepository(scope: Construct, id: string, props?: CfnPublicRepositoryProps)
\`\`\`

| **Name** | **Type** | **Description** |
| --- | --- | --- |
| [\`scope\`](#awscdkawsecrcfnpublicrepositoryparameterscope)<span title=\\"Required\\">*</span> | [\`@aws-cdk/core.Construct\`](#@aws-cdk/core.Construct) | scope in which this resource is defined. |
| [\`id\`](#awscdkawsecrcfnpublicrepositoryparameterid)<span title=\\"Required\\">*</span> | \`string\` | scoped id of the resource. |
| [\`props\`](#awscdkawsecrcfnpublicrepositoryparameterprops) | [\`@aws-cdk/aws-ecr.CfnPublicRepositoryProps\`](#@aws-cdk/aws-ecr.CfnPublicRepositoryProps) | resource properties. |

---

# \`scope\`<sup>Required</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepository.parameter.scope\\" id=\\"awscdkawsecrcfnpublicrepositoryparameterscope\\"></a>

- *Type:* [\`@aws-cdk/core.Construct\`](#@aws-cdk/core.Construct)

scope in which this resource is defined.

---

# \`id\`<sup>Required</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepository.parameter.id\\" id=\\"awscdkawsecrcfnpublicrepositoryparameterid\\"></a>

- *Type:* \`string\`

scoped id of the resource.

---

# \`props\`<sup>Optional</sup> <a name=\\"@aws-cdk/aws-ecr.CfnPublicRepository.parameter.props\\" id=\\"awscdkawsecrcfnpublicrepositoryparameterprops\\"></a>

- *Type:* [\`@aws-cdk/aws-ecr.CfnPublicRepositoryProps\`](#@aws-cdk/aws-ecr.CfnPublicRepositoryProps)

resource properties.

---
"
`;
